### DISCUSSION TITLE
IModule.isModified() always returns true in pre/post save hook

### DISCUSSION THREAD

**Discussion by kschmiechen** [SchweizerischeBundesbahnen/ch.sbb.polarion.extension.interceptor-manager#103] [General]
During my hook implementations, I figured that IModule.isModified() always returns true in pre/post save hooks, even if the structure of the document has not been changed (structure change: document newly created, work items added/removed/moved).

I reported this issue to Polarion Support but they just forwarded me to you.
I don't see how the Interceptor Manager should cause this behavior.
I can also reproduce this in the previous interceptor framework: https://extensions.polarion.com/extensions/146-work-item-action-interceptor-framework

Maybe you can reproduce (e.g., just change a custom field of a document and check the method in the manager itself) this issue and comment on the topic.
I'll then get back to Siemens with the additional info.

**Comment by grigoriev** [comment]
Hi @kschmiechen,

we are not modifying the state of the objects in the interceptor-manager.

I've made simple tests.

```java
    @Override
    public String preAction(@NotNull IPObject object) {
        @NotNull IModule module = (IModule) object;
        boolean modified = module.isModified();
        ...
    }
```

Case 1: document has no structural changes (`module.xml` has NOT been changed) -> modified is `false`
Case 2: document has any structural changes (`module.xml`) -> modified is `true`

May be it's related to the understanding of "**structural changes**". You asked for _"just change a custom field of a document"_, from my point of view this is "**structural changes**", because it changes the `module.xml` (all the fields are stored there).

**Reply by kschmiechen** [reply]
This is the documentation of `isModified()`:
`Whether this module structure (tree of IStructureNode returned by getRootNode) was modified after it was read from storage (newly created modules are always modified).`
And this is also the behavior I get outside of a save hook context.

So what did you change in Case 1?

**Reply by grigoriev** [reply]
In Case 1 I've changed the workitem inside the document.
<img width="818" alt="image" src="https://github.com/user-attachments/assets/5a5e4c08-b72a-4fa5-ab33-3b909cf23d7e">

**Reply by kschmiechen** [reply]
If it's not too much work, could you also test the result of the method in a workflow condition when only the status has changed?
There it returns false for me. In Case 2 that you already tested (in hook context) where only a field (e.g. status) has changed, it returns true.

**Reply by grigoriev** [reply]
Could you please provide me with details how to do it?

**Reply by kschmiechen** [reply]
I did some additional tests with the following script workflow condition (rename file to js).
[isModifiedTest.txt](https://github.com/user-attachments/files/17028638/isModifiedTest.txt)
It seems like the method does not return true for the status field but for all others
Steps to reproduce:
1. Setup workflow condition for arbitrary workflow action
2. Create new livedoc for the configured type
3. modify the default text and save (important step)
4. change the status using the configured workflow action -> method returns false
5. change the status back to the initial status and save
6. modify a different field than the status field and perform the workflow in the same operation -> method returns true

Seems like I did not identify this behavior earlier because my tests for the workflow conditions did not contain changes to document fields.
So your understanding is correct that changes to the custom fields are considered changes.
The only exception is the status field which is unexpected.

I'll report this back to Siemens and ask for an improvement of the documentation.
